### DISCUSSION TITLE
How to add a (changing) live DJ image to a video stream (`video.add_image` doesn’t)?

### DISCUSSION THREAD

**Discussion by Moonbase59** [savonet/liquidsoap#3177] [Q&A]
I’m trying to add a DJ image to a videostream, whenever a DJ goes live. If no DJ is live, the station logo shall be shown.

I have

```
logo_file = station_base_dir ^ "/media/videostream/niteradio-cover.png"
live_dj_image = ref(logo_file)
```
and set the `live_dj_image` from AzuraCast’s DJ image. This works, but using `video.add_image` just takes the first image encountered (station logo) as a static image (as stated in the [docs](https://www.liquidsoap.info/doc-dev/reference.html#video.add_image)) and never updates:

```
# Add station logo -- or live DJ image
s = video.add_image(x=50, y=50, width=100, height=100, file="#{live_dj_image()}", s)
```

**What would be a good (and possibly fast) way to add an image that can change?**

The DJ image links I get from AzuraCast look like these:
```
https://example.com/api/station/1/streamer/3%7C1681048534/art
```

**Comment by smimram** [comment]
The way you write it the filename is computed once. You should write
```
s = video.add_image(x=50, y=50, width=100, height=100, file=live_dj_image, s)
```
(note the difference on `file`)

**Reply by Moonbase59** [reply]
Aren’t we missing a `()` here (it’s a ref)? I already tried `…, file=live_dj_image(), s)`, but that also never changes (uses the station logo).

The log says I got the correct image:
```
      if dj.has_custom_art then
        # if we have a DJ image, use it instead of the station logo
        live_dj_image.set(dj.art)
        log("DJ IMAGE: #{live_dj_image()}")
      else
        # no image, use station logo
        live_dj_image.set(logo_file)
      end
```

```
2023/06/29 17:27:18 [lang:3] DJ IMAGE: https://example.com/api/station/1/streamer/3%7C1681048534/art
```

Using Liquidsoap 2.2.0+git@a3e2952f9.

**Reply by vitoyucepi** [reply]
1. `live_dj_image = ref(logo_file)` is reference to a string.
1. `live_dj_image` is also a function.
1. `live_dj_image()` is a string contained in the reference.
1. `"#{live_dj_image()}"` is a string inside a string interpolation.

In liquidsoap before *2.2.x* you have to write `file={!live_dj_image}`.
1. `{}` is an anonymous function.
1. `!live_dj_image` is a string contained in the reference.

When the anonymous function runs, it will extract the string.

If you write `file=!live_dj_image`, then the string will be extracted once.

**Reply by toots** [reply]
No, no `()`. With the `()` you are applying the function so it gets executed once and forever. Without it, you are passing the entire function the operator and it can execute it at will to dynamically change the image.

**Reply by toots** [reply]
@vitoyucepi said it better :-)

**Comment by Moonbase59** [comment]
Great explanation @vitoyucepi. BUT… `live_dj_image` _is_ actually a ref to a string (the filename I get from the AzuraCast API), and consequently, when using just `live_dj_image`, I get:

```
LogicException at /var/azuracast/www/src/Radio/Backend/Liquidsoap.php L247: At line 768, char 60-73: Error 5: this value has type () -> _ (inferred at line 657, char 16-30) but it should be a subtype of the type of the value at video.liq, line 58, char 109-113 string 
```

From `video.add_image`, if I understand correctly, it _does_ expect a string for `file`, right?
```
(?id : string?, ?width : {int}?, ?height : {int}?, ?x : {int}, ?y : {int},
 file : string, source(video=canvas('a), 'b)) -> source(video=canvas('a), 'b)
```

Also, `video.add_rectangle` and `video.add_text` need the `{}` and _work_:
```
# Show live DJ (if any) -- must use transparency & empty text to hide
s = video.add_rectangle(alpha={live_dj_alpha()}, color={live_dj_color()}, height=30, width=320, x=0, y=532, s)
s = video.add_text(font=font_file, size=20, x=50, y=537, color=0xebebeb, {live_dj_text()}, s)
```

The basic idea of all this is: I set all these `live_dj_*` variables whenever the (audio) metadata changes:
```
# API calls take time, let’s run this in a separate thread, so we don’t block anything
a.on_metadata(fun(m) -> thread.run({update_nowplaying(m)}))
```
so that I can use them in building the audio-video source that gets output to Icecast:
```
# Add station logo -- or live DJ image
s = video.add_image(x=50, y=50, width=100, height=100, file=live_dj_image(), s)

# Now playing area
s = video.add_rectangle(alpha=0.5, color=0x444444, height=108, width=1280, x=0, y=562, s)
s = video.add_text(font=font_file, size=40, x=50, y=572, color=0xebebeb, {nowplaying()}, s)

# Show live DJ (if any) -- must use transparency & empty text to hide
s = video.add_rectangle(alpha={live_dj_alpha()}, color={live_dj_color()}, height=30, width=320, x=0, y=532, s)
s = video.add_text(font=font_file, size=20, x=50, y=537, color=0xebebeb, {live_dj_text()}, s)

# progress bar for video track
s = video.add_rectangle(alpha=0.75, color=0xebebeb, x=0, y=666, height=4, width={int(px_per_pct*v_percent(v))}, s)

# Icecast video stream shall get its own clock, for decoupling
videostream = mksafe(buffer(s))
clock.assign_new(id="videostream", [videostream])

# reducing bitrate below b="5000k" produces horrible results
# if we leave the "b=" off, this will generate about 4000-12000 kbit/s
# so max. 3 listeners when selfhosting on 100 MBits/s (~30 MBits/s upload)
# If limiting bitrate, use "maxrate" and "bufsize" together
# (bufsize = 2 x maxrate seems to work good)
# Setting the GOP (group-of-pictures) to twice the framerate helps some decoders,
# most notably KODI. KODI 20.1 on a FireTV 4K stick also sucks above 10000k bitrates.
# It is certainly best to stream to a high bandwidth relay!
# Resonable 720p25 encoding can be had at the cost of about 190% CPU (multi-threaded).
output.icecast(%ffmpeg(format="mpegts",
  %audio(codec="ac3",channel_coupling=0),
  %video(codec="libx264",g="50",
         maxrate="8000k",bufsize="16000k",
         preset="ultrafast", tune="zerolatency")),
         id="videostream", host = "127.0.0.1", port = 8000, password = "mysecretpassword", mount = "/video.m2t", name = "Nite Radio", description = "Non-public test & evaluation server only", genre = "Various", url = "https://example.com/public/niteradio", public = false, encoding = "UTF-8", send_icy_metadata=true, videostream)
```

**Reply by vitoyucepi** [reply]
I discovered that the `file` must be in the format of a `string`, not `{string}`.
I have submitted a feature request #3178.

**Reply by Moonbase59** [reply]
That’s what I feared… Thanks!

**Comment by Moonbase59** [comment]
@vitoyucepi I’ll save your great explanation into my list of "Vito notes"—you really have a talent for explaining things!

Nevertheless, I’ll unmark it as an answer, since it doesn’t solve the original problem. I fear we might have an oversight or actually a bug with `video.add_image`.

**Comment by Moonbase59** [comment]
@vitoyucepi, @toots: For my understanding (talking about LS 2.2.x): So _that_ was the main reason switching from `!var` notation to using `var()`?

If I understand the above correctly, using code like
```
nowplaying = ref(something)
…
s = video.add_text(font=font_file, size=40, x=50, y=572, color=0xebebeb, {nowplaying()}, s)
```
is inefficient, because I would extract the referenced _value_, only to then put it into an anonymous function again, by using `{nowplaying()}`, and I _should_ use `nowplaying` instead because a ref now _behaves_ like a function? Which would make the above a much easier and better readable
```
s = video.add_text(font=font_file, size=40, x=50, y=572, color=0xebebeb, nowplaying, s)
```

Did I understand that correctly?

**Reply by toots** [reply]
The main reason for the switch is that `!var`, the old notation for reading a reference, is super confusing to most people. That's an OCaml syntax but, in most other languages, `!var` is the logical negation, which, for us, is `not var`.

Then it turned out that making reference functions was also less to learn (it's just a function with a setter!) and fit very well into the getter use-cases.

Lastly, we've kept `var := 123` because this one does make sense. It became a syntactic sugar for `var.set(123)` which, in turn allowed to unify settings with references and write things like:
```
log.level := 4
```
which I think are easy to understand to most people

**Reply by toots** [reply]
Ha also, if you write `{now_playing()}` you are creating this function: `fun () -> now_playing()` which is a function that executes `now_playing` each time it is called. This is the same as simply passing `now_playing` (it could potentially be different if `now_playing` had optional arguments).

**Reply by Moonbase59** [reply]
Thanks for confirming my thoughts! I actually now streamlined my code a little, getting rid of all these `{func()}` and replacing them by `func` where appropriate. And you’re right, for functions with arguments I actually need to use `{func(x)}`.

Any plans on implementing #3178?
